src/server/pkg/hashtree/db.go

```diff
@@ -0,0 +1,1703 @@
+}
+
+// Writer can write a serialized hashtree from a sequence of merge nodes.
+type Writer struct {
+	pbw    pbutil.Writer
+	size   uint64
+	idxs   []*Index
+	offset uint64
+}
+
+// NewWriter creates a new hashtree writer.
+func NewWriter(w io.Writer) *Writer {
+	return &Writer{
+		pbw: pbutil.NewWriter(w),
+	}
+}
+
+// Write writes the next merge node.
+func (w *Writer) Write(n *MergeNode) error {
+	// Marshal node if it was merged
+	if n.nodeProto != nil {
+		var err error
+		n.v, err = n.nodeProto.Marshal()
+		if err != nil {
+			return err
+		}
+	}
+	// Get size info from root node
+	if bytes.Equal(n.k, nullByte) {
+		if n.nodeProto == nil {
+			n.nodeProto = &NodeProto{}
+			if err := n.nodeProto.Unmarshal(n.v); err != nil {
+				return err
+			}
+		}
+		w.size = uint64(n.nodeProto.SubtreeSize)
+	}
+	// Write index for every index size bytes
+	if w.offset > uint64(len(w.idxs)+1)*IndexSize {
+		w.idxs = append(w.idxs, &Index{
+			K:      n.k,
+			Offset: w.offset,
+		})
+	}
+	b, err := w.pbw.WriteBytes(n.k)
+	if err != nil {
+		return err
+	}
+	w.offset += uint64(b)
+	b, err = w.pbw.WriteBytes(n.v)
+	if err != nil {
+		return err
+	}
+	w.offset += uint64(b)
+	return nil
+}
+
+// Copy copies a hashtree reader in a writer.
+func (w *Writer) Copy(r *Reader) error {
+	for {
+		n, err := r.Read()
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+		if err := w.Write(n); err != nil {
+			return err
+		}
+	}
+}
+
+// Index returns the index for a hashtree writer.
+func (w *Writer) Index() ([]byte, error) {
+	buf := &bytes.Buffer{}
+	pbw := pbutil.NewWriter(buf)
+	for _, idx := range w.idxs {
+		if _, err := pbw.Write(idx); err != nil {
+			return nil, err
+		}
+	}
+	return buf.Bytes(), nil
+}
+
+// GetRangeFromIndex returns a subtree byte range in a serialized hashtree based on a passed in prefix.
+func GetRangeFromIndex(r io.Reader, prefix string) (uint64, uint64, error) {
+	prefix = clean(prefix)
+	pbr := pbutil.NewReader(r)
+	idx := &Index{}
+	k := b(prefix)
+	var lower, upper uint64
+	iter := func(f func(int) bool) error {
+		for {
+			if err := pbr.Read(idx); err != nil {
+				if err == io.EOF {
+					break
+				}
+				return err
+			}
+			var cmp int
+			if len(k) < len(idx.K) {
+				cmp = bytes.Compare(k, idx.K[:len(k)])
+			} else {
+				cmp = bytes.Compare(k[:len(idx.K)], idx.K)
+			}
+			if f(cmp) {
+				break
+			}
+		}
+		return nil
+	}
+	low := func(cmp int) bool {
+		if cmp > 0 {
+			lower = idx.Offset
+			return false
+		} else if cmp < 0 {
+			// Handles the case where a prefix fits within one range
+			upper = idx.Offset
+		}
+		return true
+	}
+	up := func(cmp int) bool {
+		if cmp < 0 {
+			upper = idx.Offset
+			return true
+		}
+		return false
+	}
+	// Find lower
+	iter(low)
+	// Find upper
+	if upper <= 0 {
+		iter(up)
+	}
+	// Handles the case when at the end of the indexes
+	if upper <= 0 {
+		return lower, 0, nil
+	}
+	// Return offset and size
+	return lower, upper - lower, nil
+}
```

Comment:
I agree, I struggled trying to make this more readable, but did not have too much luck. There is probably a cleaner way to do this, but it will probably require more time to think it through. At a high level I think it is pretty straightforward in that I iterate until I find the lower bound, then iterate until I find the upper bound. The parts that are confusing are the multiple edges cases. I think the comments I have explain the edge cases, but I guess I can see if I can add a little bit more.